pkg/metrics/collector/kernel_metrics.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	tcpTimeoutDesc = prometheus.NewDesc(
+		"tcp_timeout_rehash_count",
+		"Tcp timeout rehash, collected from nstat",
+		nil, nil,
+	)
+	tcpDuplicateDesc = prometheus.NewDesc(
+		"tcp_duplicate_rehash_count",
+		"Tcp duplicate rehash, collected from nstat",
+		nil, nil,
+	)
+	segmentsReceivedDesc = prometheus.NewDesc(
+		"tcp_segments_received_count",
+		"TCP segments received on the node, collected from netstat -s",
+		nil, nil,
+	)
+	segmentsSentDesc = prometheus.NewDesc(
+		"tcp_segments_sent_count",
+		"TCP segments sent from node, collected from netstat -s",
+		nil, nil,
+	)
+	segmentsRetransmittedDesc = prometheus.NewDesc(
+		"tcp_segments_retransmitted_count",
+		"TCP segments retransmitted on the node, collected from netstat -s",
+		nil, nil,
+	)
+)
+
+type kernelStatCollector struct {
+}
+
+func init() {
+	registerCollector("kernel_metrics", NewKernelStatCollector)
+}
+
+func NewKernelStatCollector() (Collector, error) {
+	return &kernelStatCollector{}, nil
+}
+
+// parseNetstatVals parses the contents of /proc/net/netstat
+// looking for lines such as:
+//
+// TcpExt: SyncookiesSent RcvPruned OfoPruned OutOfWindowIcmps LockDroppedIcmps TcpTimeoutRehash TcpDuplicateDataRehash TCPDSACKRecvSegs TCPDSACKIgnoredDubious
+// TcpExt: 0 0 0 0 0 440 99 21800 3
+// and populates a map with all the parsed values where the keys are
+// the labels in the first line, separated by spaces and second line contains the values,
+// indexed accordingly
+func parseNetstatVals(output string) map[string]uint64 {
+	stats := make(map[string]uint64)
+	splitFile := strings.Split(output, "\n")
+	for i, stat := range splitFile {
+
+		if !strings.Contains(stat, "TcpExt:") {
+			continue
+		}
+		keyList := strings.Fields(stat)
+		valList := strings.Fields(splitFile[i+1])
```

Comment:
Add a TODO here to look for the next instance of TcpExt: instead of assuming it's the immediately following line.

No need to change it, just leave a TODO here as a breadcrumb if anybody comes looking later.

Reply:
Done